Ensure parameters are in scope when converting parameter/return types to type nodes

jakebailey opened the pull request:
Fixes #48783

jakebailey commented on tests/baselines/reference/jsDeclarationsParameterTagReusesInputNodeInEmit1.js:
```diff
@@ -86,7 +86,7 @@ declare const couldntThinkOfAny: {};
  * @returns {InstanceType<BaseFactory["Base"]>}
  */
-declare function test(base: InstanceType<BaseFactory["Base"]>): InstanceType<BaseFactory["Base"]>;
+declare function test(base: InstanceType<BaseFactory["Base"]>): {};
```
This isn't quite right; will look.

jakebailey commented on tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespaceCjs.js:
```diff
@@ -96,4 +96,69 @@ export const testFnTypes: {
 import { myTypes } from "./file.js";
+
+
+//// [DtsFileErrors]
```
This shouldn't have been turned into an error, will check.

jakebailey commented on tests/baselines/reference/jsDeclarationsFunctions.js:
```diff
@@ -173,6 +173,6 @@ export function g(a: {
     x: string;
 }, b: {
-    y: typeof b;
+    y: typeof import(".").b;
```
This is suspicious; the code above says defines param `b`'s type as `{y: typeof b}` , which is self referential and surely wrong.

jakebailey commented:
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this

jakebailey commented on tests/baselines/reference/blockScopedVariablesUseBeforeDef.types:
```diff
@@ -70,12 +70,12 @@ function foo5() {
 >x : () => () => any
 >() => y : () => () => any
->y : () => () => any
+>y : () => any
```
This comes from:

```
function foo5() {
    let x = () => y;
    let y = () => x;
}
```

So, it's just where this infinite recursion ends; the tooltip says something else entirely too.

jakebailey commented:
@typescript-bot pack this

sandersn commented:
I haven't reviewed the code yet, but I am weirded out by the whitespace changes -- which are admittedly the same as in #37444. But why insert 3 more spaces in, say, `left` and `right` below? It doesn't correspond to initial 4-space indent -- `top` is indented 8 spaces -- and `left` and `right` are only preceded by one space in the source.

```ts
module A {

    class Point {
        constructor(public x: number, public y: number) { }
    }

    export var UnitSquare : {
        top: { left: Point, right: Point },
        bottom: { left: Point, right: Point }
    } = null;
}
```

jakebailey commented:
I think before it was creating an all new type node and in doing so put the code onto one line. But after, it's reusing the node, and so it has the spacing as the original source. The baselines just remove the newlines so it looks awkwardly spaced.

I'd test on the above pack-this, but it looks like typescript-staging is down :(

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -5880,6 +5880,24 @@ namespace ts {
+                // only be typed via JSDoc, but JSDoc annotations appear outside of the function and therefore are not in
+                // the function's scope, unlike true type annotations.
+                const saveEnclosingDeclaration = context.enclosingDeclaration;
+                if (signature.declaration && !isInJSFile(signature.declaration) && saveEnclosingDeclaration && findAncestor(signature.declaration, (n) => n === saveEnclosingDeclaration)) {
```
```suggestion
                if (signature.declaration && !isInJSFile(signature.declaration) && saveEnclosingDeclaration && findAncestor(signature.declaration, n => n === saveEnclosingDeclaration)) {
```

jakebailey replied:
ugh github never lets me apply these

weswigham commented on src/compiler/checker.ts:
```diff
@@ -5880,6 +5880,24 @@ namespace ts {
+                // only be typed via JSDoc, but JSDoc annotations appear outside of the function and therefore are not in
+                // the function's scope, unlike true type annotations.
+                const saveEnclosingDeclaration = context.enclosingDeclaration;
+                if (signature.declaration && !isInJSFile(signature.declaration) && saveEnclosingDeclaration && findAncestor(signature.declaration, n => n === saveEnclosingDeclaration)) {
```
I don't really like using the declaration from the signature over the context passed in - it may be in an entirely different file from the context where the signature is being emitted to, so will have potentially entirely incorrect results for things-that-aren't-immediately-parameters/type-parameters-of-the-signature. Instead we probably need to track the parameters we're serializing as we recur through multiple signatures - we already do this for type parameters in some cases. Similarly, we manufacture a fake node to use as a context in some places in the JS declaration builder that handles blocking/preserving scope variables like you'd want, since the real node doesn't exist (like how a `module.export.x = expr` needs a fake `module.exports` namespace declaration - passing in better initial contexts like this also sometimes an option.

jakebailey replied:
I was also concerned about this (it felt wrong entirely ignoring the context), but wasn't sure what to do instead. I'll try to come up something more like those other methods.

jakebailey replied:
> it may be in an entirely different file from the context where the signature is being emitted to, so will have potentially entirely incorrect results for things-that-aren't-immediately-parameters/type-parameters-of-the-signature (meaning it may allow references to locals in a separate file as though they're locals in the emitted file!)

Actually, just to clarify this, it was my impression that the ancestor check I've done here avoided that problem because it verifies that the current signature is a descendent of the current enclosing declaration.

This behavior is already not legal for declarations in other files because this involves `typeof` on the function itself. You can't directly do a `typeof` for a parameter that comes from an entirely different file (or even a different scope). The `findAncestor` check is what fixed that bug when I was working on this particular fix as there were some tests that did exhibit the bad behavior.

weswigham replied:
If the function comes from within another function, but is being serialized in the top level scope of a js file, the scopes in between are still inaccessible. So some scope structure like
```
function f() {
  return g;
  /** @returns {typeof h} */
  function g() { return h; }
  function h() {}
}

module.exports = f;
```
Though probably with parameter references to inner functions in some way.

jakebailey replied:
I'm not sure if this is quite what you were suggesting, but the following doesn't work with my PR:

```ts
export function foo() {
    const bar = (isNan: typeof globalThis.isNaN): typeof globalThis.isNaN => isNan

    function baz() {
        return bar;
    }

    return baz;
}
```

[Playground Link](https://www.staging-typescript.org/play?ts=4.8.0-pr-49627-10&ssl=11&ssc=2&pln=3&pc=1#code/FAehAIHUCcHsDsDmAucBnAFrArgGwCboCGAnuAAYAuJADgKawBm4iusARkbgCoYCWaAHQCAckRHlwdAG51oJAO4Y5dcH3jhKy8PkGU0wOgA8asaJXCNs8AMaU+CS7FgAKAJTgA3sHC-wNhDQLTmhwAF5wF1EieFRqeiYWNk4efiFokTc42gZmVg4uXgFhNDERcIA+NVKY4B8-K1t7R04AL3cver8-aDpKbGgNEIBuLvAAXzru3v7B8DbRycMTMwsA+CDwIgBGcMiM7IS85MK0krKszRzE-JSi9JrysKqM0eNTc39AiyIAJj2oo9DrkkgVUsUMgAaeZEaAAfmBNxO4IeF0RxzB93O4kqMNCcLh1TKbxWn3WmyIAGYAQdwPBsABbdhyaEhdGgu5nDKXeIg26nCGPXEjZYfNbfLYAFhpQLpjOZ0B51wxnMFZVx-JR2JEwyAA)

I'm not sure I can actually produce a counterexample in JS, though; JSDoc is exempt from this PR just because the scope accessible in a JSDoc comment is outside the declaration, not inside of it.

jakebailey replied:
Hm, probably not what you meant; I think this just fails because the enclosing declaration is not a parent of the signature (which this PR used to work around things).

weswigham commented on src/compiler/transformers/declarations.ts:
```diff
@@ -677,15 +677,20 @@ namespace ts {
-                || isIndexSignatureDeclaration(node)
-                || isMappedTypeNode(node);
+        function isEnclosingDeclaration(_node: Node) {
+            // TODO(jakebailey): Clean this up. If we look at the way that the "types"
```
Just as an FYI, this list is the nodes our old declaration emitter captured as enclosing declarations while recurring, not many other reasons for it.

jakebailey replied:
Just to clarify, do you agree with removing this entirely or the opposite?

This is just a leftover now, I'll probably not touch this (I needed it in a previous iteration but haven't yet removed it, this branch is big WIP right now)

weswigham replied:
There's no documented logic for why it's only *these* nodes which are/were captured - these are simply the only ones that were way back when the declaration emitter was string-based.

jakebailey commented on tests/baselines/reference/leaveOptionalParameterAsWritten.js:
```diff
@@ -32,6 +32,4 @@ declare global {
 }
 //// [c.d.ts]
-declare type Foo = teams.calling.Foo;
-export declare const bar: (p?: Foo) => void;
```
This is a case that I'm pretty sure only worked because of a fluke. It's desirable, but when we get the type from the annotation, we get the interface type `Foo` in another file (the symbol is correctly resolved), but then this gets handed down and the info about it being an alias is completely lost. I think this only worked due to assumptions about the enclosing declaration that now no longer hold.

jakebailey commented:
I've gotten this PR pretty far but I'm going to stop working on it for a while because it's getting too frustrating. It seems like this part of the code is really fragile when it comes to what it considers to be reusable in ways that I can't seem to keep in my head at once. If fix one thing and break something I fixed earlier, because I forgot what I did and ended up undoing it, etc.

The whole enclosing declaration thing in general weirds me out; it doesn't seem consistent at all which enclosing declaration is used in what scenario, and yet the types baseline code just always uses the parent node, which would lead me to believe that it doesn't need to even exist at all, either, but I'm sure I'm missing something.

weswigham commented:
To be fair, I think the primary reason they're different is that the declaration emitter is eliding or rewriting a lot of scopes, while the types baselines aren't.

jakebailey commented:
@typescript-bot perf test this

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -7139,6 +7139,42 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                    factory.createBlock([]),
+                );
+
+                // TODO(jakebailey): Can we skip this if there are no parameters?
+                // This seems to bring back desirable node reuse, but I'd want to see
+                // functions with parameters fixed too.
```
The answer here is yes, but it does come at a const of inconsistency in emit because going from zero parameters to one (or vice versa) may change the emitted return type entirely, like:

```diff
+>d2 : { foo(): {    x: number;}; }
->d2 : { foo(): { x: number; }; }
```

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -7139,6 +7139,42 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            // JSDoc, which are always outside the function declaration, so are not in the parameter scope.
+            const saveEnclosingDeclaration = context.enclosingDeclaration;
+            if (saveEnclosingDeclaration && signature.declaration && signature.declaration !== saveEnclosingDeclaration && !isInJSFile(signature.declaration)) {
+                const fakeScope = parseNodeFactory.createFunctionExpression(
```
While this method does work, it has the downside of introducing a synthetic node into the declaration chain, and so when this happens, the emit suffers. I'm not sure of a better way; `isSymbolAccessible` currently seems to only use the enclosing declaration and its parents, so I don't know how to get this info into its scope.

jakebailey commented:
@typescript-bot perf test this

jakebailey commented:
Cool, seems like no performance change. I was worried. I think this is back and probably ready for review.

I've managed to fix most cases where we were unable to reuse the existing node, but a few remain. It's really funky because given:

```ts
interface I {
    f: <T>(a: { /*1*/T: number }, b: T) => T;
    g: (a: { /*2*/a: number }) => void;
}
```

We walk down to `a` and then find that there's another `a` in scope via the enclosing declaration and so we don't think we can reuse the node. Except, there's no way to actually refer to that inner `a` so it's just waste. I'm not quire sure how to detect this case but I can't seem to actually find an example where this becomes harmful (even though the code says that it might affect quickinfo; testing says no).

jakebailey commented:
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot pack this

jakebailey commented:
Forgive the force push; this had so many commits that I wanted to still be able to see the test diffs.

I've updated the logic to share the same node within the same chain instead of repeating it multiple times; I think it's about as fast as it can get.

I also dropped the added property on the context; it turns out that wasn't safe because it's technically possible that another part of the code temporarily changes the enclosing declaration, but we could then accidentally inject new variables into a different chain. Now, we just walk up the ancestors to find an existing one, creating one if needed.

jakebailey commented on tests/baselines/reference/declarationEmitGlobalThisPreserved.js:
```diff
@@ -168,125 +168,3 @@ export declare class AsClassProperty {
 export type AsFunctionType = (isNaN: typeof globalThis.isNaN) => typeof globalThis.isNaN;
-
-
-//// [DtsFileErrors]
```
Note the the errors going away here; I constructed these tests such that they will emit a d.ts error when they're wrong, rather than us having to figure out if the output looks right or not.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -7276,6 +7276,82 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                // It'd likely be better to store this somewhere else for isSymbolAccessible, but
+                // since that API _only_ uses the enclosing declaration (and its parents), this is
+                // seems like the best way to inject names into that search process.
+                const existingFakeScope = findAncestor(context.enclosingDeclaration, node => !!getNodeLinks(node).fakeScopeForSignatureDeclaration);
```
I struggle to think of when you'd have another synthetic scope set as the child scope of an existing fake scope now. Maybe this can just be `!!getNodeLinks(context.enclosingDeclaration).fakeScopeForSignatureDeclaration` ?

jakebailey replied:
The case I'm worried about is the other places we reassign `context.enclosingDeclaration` temporarily, which happens in `serializeTypeAlias`, `sanitizeJSDocImplements`, `serializeAsClass`, `addPropertyToElementList`, etc.

I'd have to come up with an example of this, though.

jakebailey replied:
Okay, it does work to only ever check the most immediate enclosing declaration, at least in the the current tests. I would have thought it'd break, but examining things again, I can't actually figure out a counterexample here, so I'll switch things again.

jakebailey replied:
And to clarify, I did add an assert that verifies that there aren't more than 1 fake scope in a given chain, and it never fired, so that's proof enough I suppose.

jakebailey replied:
Switched to only check the current enclosingDeclaration.

sandersn replied:
Just started reviewing this -- does this mean the comment block above also needs to be simplified?

jakebailey replied:
Yes, let me fix the wording. Most of it still applies but now there's only one fake scope.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -7996,13 +8073,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
 
+        function getEnclosingDeclarationIgnoringFakeScope(enclosingDeclaration: Node) {
+            return findAncestor(enclosingDeclaration, n => !getNodeLinks(n).fakeScopeForSignatureDeclaration);
```
Likewise, I think this should only need to check the immediate `enclosingDeclaration` now.

jakebailey replied:
Could be, though if that's the case, it'll only walk one node upward so I'm not sure I want to write it less conveniently...

jakebailey replied:
Switched this to a simple ternary.

jakebailey commented:
@typescript-bot perf test faster

sandersn commented on src/compiler/checker.ts:
```diff
@@ -7276,6 +7276,86 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                && signature.declaration
+                && signature.declaration !== context.enclosingDeclaration
+                && !isInJSFile(signature.declaration)
+                && some(expandedParams)
```
getExpandedParameters looks like it always returns an array of length 1, at least unless the rest type is somehow a 0-constituent union, but I think those are always impossible.

If the check is there just because the types could allow an empty array, maybe it should be an assert in the body of the `if` instead.

jakebailey replied:
Check above, and you'll see:

```ts
const expandedParams = getExpandedParameters(signature, /*skipUnionExpanding*/ true)[0];
```

jakebailey replied:
(Later I might see if we can avoid constructing a bunch of random arrays we don't need...

sandersn reviewed (commented):
I don't think I understand the overall flow of creating a signature declaration enough to properly sign off, but I did have a couple of questions after reading the code.
